src/category_theory/limits/presheaf.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.adjunction
+import category_theory.elements
+import category_theory.limits.functor_category
+import category_theory.limits.shapes
+import category_theory.limits.types
+
+/-!
+# Colimit of representables
+
+This file constructs an adjunction between `(Cᵒᵖ ⥤ Type u)` and `ℰ` given a functor `A : C ⥤ ℰ`,
+where the right adjoint sends `(E : ℰ)` to `c ↦ (A.obj c ⟶ E)` (provided `ℰ` has colimits).
+
+This adjunction is used to show that every presheaf is a colimit of representables.
+
+Further, the left adjoint `colimit_adj.L : (Cᵒᵖ ⥤ Type u) ⥤ ℰ` satisfies `yoneda ⋙ L ≅ A`, that
+is, an extension of `A : C ⥤ ℰ` to `(Cᵒᵖ ⥤ Type u) ⥤ ℰ` through `yoneda : C ⥤ Cᵒᵖ ⥤ Type u`.
+TODO: Show `colimit_adj.L` is unique amongst cocontinuous functors with this property.
+
+## Tags
+colimit, representable, presheaf
+-/
+
+namespace category_theory
+
+noncomputable theory
+
+open category limits
+universes u₁ u₂
+
+variables {C : Type u₁} [small_category C]
+variables {ℰ : Type u₂} [category.{u₁} ℰ]
+variable (A : C ⥤ ℰ)
+
+namespace colimit_adj
+
+/--
+The functor taking `(E : ℰ) (c : Cᵒᵖ)` to the homset `(A.obj C ⟶ E)`. It is shown in `L_adjunction`
+that this functor has a left adjoint (provided `E` has colimits) given by taking colimits over
+categories of elements.
+In the case where `ℰ = Cᵒᵖ ⥤ Type u` and `A = yoneda`, this functor is isomorphic to the identity.
+
+Defined as in [MM92], Chapter I, Section 5, Theorem 2.
+-/
+@[simps {rhs_md := semireducible}]
+def restricted_yoneda : ℰ ⥤ (Cᵒᵖ ⥤ Type u₁) :=
+yoneda ⋙ (whiskering_left _ _ (Type u₁)).obj (functor.op A)
+
+/--
+The functor `restricted_yoneda` is isomorphic to the identity functor if evaluated at the
+yoneda embedding.
+-/
+def right_is_id : restricted_yoneda (yoneda : C ⥤ Cᵒᵖ ⥤ Type u₁) ≅ 𝟭 _ :=
+nat_iso.of_components
+(λ P, nat_iso.of_components (λ X, yoneda_sections_small X.unop _)
+  (λ X Y f, funext $ λ x,
+  begin
+    dsimp [ulift_trivial, yoneda_lemma],
+    rw ← functor_to_types.naturality _ _ x f (𝟙 _),
+    dsimp,
+    simp,
+  end))
+(λ _ _ _, rfl)
+
+/--
+(Implementation). The equivalence of homsets which helps construct the left adjoint to
+`colimit_adj.restricted_yoneda`.
+It is shown in `Le'_natural` that this is a natural bijection.
+-/
+def Le' (P : Cᵒᵖ ⥤ Type u₁) (E : ℰ) {c : cocone ((category_of_elements.π P).left_op ⋙ A)}
```

Comment:
Should this be private, or should it get a better name?

Reply:
I'm happy with either choice. Maybe `restrict_yoneda_hom_equiv`? (and +`_natural` for the below)